salt/cleanup/remove-post-orchestration.sls

```diff
@@ -0,0 +1,33 @@
+include:
+  - kubectl-config
+
+{%- set target = salt['pillar.get']('target') %}
+{%- set nodename = salt.caasp_net.get_nodename(host=target) %}
+
+###############
+# k8s cluster
+###############
+
+{%- set k8s_nodes = salt['mine.get']('roles:(kube-master|kube-minion|etcd)', 'nodename', 'grain_pcre').keys() %}
+{%- if target in k8s_nodes %}
+
+{%- from '_macros/kubectl.jinja' import kubectl with context %}
+
+{{ kubectl("remove-node",
+           "delete node " + nodename,
+           check_cmd="/bin/true") }}
```

Comment:
`etcd` removal hasn't got this `check_cmd`. On one side I'd like to reuse this `sls` from the force removal orchestration, but I think in this regular `removal` orchestration, we should fail if this fails (or removing the etcd member), while on the force removal orchestration we shouldn't.

Reply:
etcd removal _must_ succeed: otherwise, the etcd cluster will be broken. If you ignore that you could leave a cluster that is unstable and unusable.

Reply:
My main point being: why not remove `check_cmd` from here too?

Reply:
Because removing a node from the API server is not so critical, so we don't want to abort the whole orchestration if this fails. In the worst case, the node will be kept in kubernetes but it will appear as not ready.

Reply:
IMO this doesn't look good from a product point of view. We should only succeed if all operations here did work as expected. Having leftovers is something we should avoid (I know it's minimal, but still, looks like a leftover from a customer point of view). If this fails IMO the orchestration should not succeed. If this orchestration does not succeed, the customer can trigger a force remove afterwards, that (as I was trying to explain in https://github.com/kubic-project/salt/pull/463/) will try this from all possible machines (`roles:kube=*`). I at least wouldn't like to see as a customer that the removal "went fine" (all green), and I can still see the removed node in `kubectl get nodes`.

Reply:
We have been doing this for a long time (this code has been moved here from the `kubelet/init.sls` file), because we decided long ago it was a good idea to go ahead even if the `kubectl delete node` fails. And I still think it is the better solution, because users should not be forced to solve any problem in their clusters with a forced removal...

Reply:
If you mean:

- [`kubelet/update-post-start-services.sls`](https://github.com/kubic-project/salt/blob/7a3c61f05ea09f5b938caa90d0b09d04a2f09662/salt/kubelet/update-post-start-services.sls): that's contained in the migration process, and it needs to be idempotent.
- [`kubelet/stop.sls`](https://github.com/kubic-project/salt/blob/master/salt/kubelet/stop.sls#L33-L39) (no `check_cmd = /bin/true`)

Reply:
So leaving aside that we haven't been doing that for a long time; if something fails, we should not silently succeed. That's my 2 cents.

Reply:
I would prefer to not silently swallow the error. However I would like to confirm what I understood by reading the comments and the code.

The problem arises when we are removing an etcd node and something goes wrong. Am I right?

If the error on kubectl delete is ignored:

  * we have better chances to have a working etcd cluster, that's because the orchestration will find another node to assign the `etcd` role and the cluster will be healthy.
  * the user won't be notified that something went wrong. Even if we had the perfect UI/UX to show logs there's no way we can catch that because we simply ignored the error.
  * `kubectl get nodes` will still report the node. However kubernetes will realize the node is no longer reachable and will act accordingly (if that's a worker node no more workloads will be scheduled there).

If we don't swallow the error:

  * The orchestration won't assign another etcd node to the cluster. That means the etcd cluster might be close to breakage (eg: only 2 out of 3 nodes are running).
  * The node to be removed will still be shown into the UI.
  * The user is aware of the failure. One day it will also be easier for him to understand what went wrong, now he would have to look into the logs/open a ticket.
  * Can the user trigger another clean removal? What would happen then?
  * Is the user informed about the potential issue of the etcd cluster (eg: 2/3 nodes up)?
  * Is there a way for the user to grow the etcd cluster? Let's assume he leaves the failed to be removed node there and he adds a new worker/master node?

Reply:
> The problem arises when we are removing an etcd node and something goes wrong. Am I right?

No, this is just for `kubectl delete node` failing. I think we shoud ignore it and continue with the removal orchestration (it will result in a node `NotReady` showing in Kubernetes) while @ereslibre thinks we should abort the orchestration.